### PR TITLE
feat: AzureSQL masking pipeline

### PR DESCRIPTION
**Pull request by sumeetdas-dlpx**
<details open>
<summary><h2> Problem </h2></summary>

We want to make a DCS for Azure template that mask data available in Azure SQL DB.

</details>

<details open>
<summary><h2> Solution </h2></summary>

Add templates for DCS for Azure Azure SQL to Azure SQL for masking.

Fixed an issue which caused the Sink activity to fail when inserting data into a table having an `IDENTITY` column. (DAAS-9463)

</details>

<details>
<summary><h2> Testing Done </h2></summary>

Followed the build instructions and imported the template into an existing data factory.

Successfully masked a table with all supported data types and an identity column and inserted the masked data into a sink schema table using the Azure SQL masking pipeline.

**Note**: ~Currently facing an issue masking a table without identity column. Created DAAS-9463 ticket for the same.~ This issue is fixed in the latest commit.

### Table Descriptions 
Ran tests on tables – TestDataTypeTable and Employee.  

#### TestDataTypeTable 

TestDataTypeTable has columns corresponding to all data types in Azure SQL. Designed appropriate masking algorithms for each column and ran the masking pipeline successfully each when FAIL_ON_NON_CONFORMANT_ERROR is set to true and then to false. Table schema used:
```sql
CREATE TABLE {schema_name}.TestDataTypeTable (
            TinyIntColumn TINYINT,
            SmallIntColumn SMALLINT,
            IntegerColumn INT,
            BigIntColumn BIGINT,
            BitColumn BIT,
            DecimalColumn DECIMAL(18, 4),
            NumericColumn NUMERIC(18, 4),
            MoneyColumn MONEY,
            SmallMoneyColumn SMALLMONEY,
            FloatColumn FLOAT,
            RealColumn REAL,
            DateColumn DATE,
            TimeColumn TIME(3),
            DateTime2Column DATETIME2(3),
            DateTimeOffsetColumn DATETIMEOFFSET(3),
            DateTimeColumn DATETIME,
            SmallDateTimeColumn SMALLDATETIME,
            CharColumn CHAR(100),
            VarcharColumn VARCHAR(255),
            TextColumn TEXT,
            NCharColumn NCHAR(100),
            NVarcharColumn NVARCHAR(255),
            NTextColumn NTEXT,
            BinaryColumn BINARY(100),
            VarBinaryColumn VARBINARY(255),
            ImageColumn IMAGE,
            GeographyColumn GEOGRAPHY,
            GeometryColumn GEOMETRY,
            JsonColumn JSON,
            UniqueIdentifierColumn UNIQUEIDENTIFIER,
            XmlColumn XML
        )
``` 

 
For testing FAIL_ON_NON_CONFORMANT_ERROR =  false case, I assigned the DateShiftDiscrete algorithm to CharColumn (which contains random text but no dates). Verified that the values in CharColumn of sink schema’s table were the same as that of source schema’s table, which is expected since when we tell the masking API to ignore non-conformant errors, it simply returns the unmasked data in the response. 

#### Employee 

The purpose of masking Employee table was to test sink activity when a column is defined as Identity. The table has RowID column that is defined as “INT IDENTITY(1, 1)”, meaning that it can hold integer values and that the value is automatically incremented by 1 and inserted into the RowID column. The issue with IDENTITY columns is that we cannot insert values in them manually – only the DB can insert values. To overcome this, we must enable IDENTITY_INSERT property. 

Employee table schema:
```sql
CREATE TABLE {schema_name}.Employee (
            RowID INT UNIQUE,
            EmployeeID INT,
            FirstName VARCHAR(255),
            LastName NVARCHAR(255),
            DateOfBirth DATE,
            StartDateTime DATETIME,
            ProfileImage VARBINARY(255),
            Salary MONEY,
            IsActive BIT
        )
```
 

### Results 

Ran masking pipeline with FAIL_ON_NON_CONFORMANT_ERROR set to true and it was successful. It masked both TestDataTypeTable and Employee tables without any error. 

Screenshot 2024-11-05 at 12 52 35 AM

Screenshot 2024-11-05 at 1 02 42 AM 

Test results zip file: [fail_non_conformant_true.zip](https://github.com/user-attachments/files/17628441/fail_non_conformant_true.zip)

 
Ran masking pipeline with FAIL_ON_NON_CONFORMANT_ERROR set to true and it was successful. It masked both TestDataTypeTable and Employee tables without any error. Verified that the CharColumn values (which had non-conformant errors) were the same in both source and sink schemas. 
 
Screenshot 2024-11-05 at 1 41 44 AM

Screenshot 2024-11-05 at 1 48 46 AM
 
 
Test results zip file: [fail_non_conformant_false.zip](https://github.com/user-attachments/files/17628437/fail_non_conformant_false.zip)

</details>

### CODE REVIEW

File: dcsazure_AzureSQL_to_AzureSQL_mask_pl/dcsazure_AzureSQL_to_AzureSQL_mask_pl.json [hunk 1]
```diff
@@ -0,0 +1,1889 @@
...
+                              "DF_SOURCE_SCHEMA": {
+                                "value": "'@{item().source_schema}'",
+                                "type": "Expression"
+                              },
+                              "DF_SOURCE_TABLE": {
+                                "value": "'@{item().source_table}'",
+                                "type": "Expression"
+                              },
+                              "DF_COLUMN_WIDTH_ESTIMATE": "1000",
+                              "DF_FILTER_KEY": "''",
```
Comment (ankurs-delphix) [inline dcsazure_AzureSQL_to_AzureSQL_mask_pl/dcsazure_AzureSQL_to_AzureSQL_mask_pl.json:286]:
```suggestion
                              "DF_FILTER_KEY": {
                                  "value": "'@{item().filter_alias}'",
                                  "type": "Expression"
                              },
```

**Review by ankurs-delphix — COMMENTED** [review]
Set this in the preSQL for both sink (filtered and non filtered)

```bash
concat('SET IDENTITY_INSERT', ' ', $DF_SINK_SCHEMA, '.', $DF_SINK_TABLE, ' ', 'ON')
```

```bash
concat('SET IDENTITY_INSERT', ' ', $DF_SINK_SCHEMA, '.', $DF_SINK_TABLE, ' ', 'OFF')
```

Syntax is 
```
SET IDENTITY_INSERT <SCHEMA>.<TABLE_NAME> <ON|OFF>
```

PS: I've tested this with table that has identity enabled (Auto incrementing primary key). Request you to test your flow with this enabled. 

You might need to add directly in the SINK settings and reexport the template.

File: dcsazure_AzureSQL_to_AzureSQL_mask_pl/dcsazure_AzureSQL_to_AzureSQL_mask_pl.json (hunk 1 above)
Comment (ankurs-delphix) [reply dcsazure_AzureSQL_to_AzureSQL_mask_pl/dcsazure_AzureSQL_to_AzureSQL_mask_pl.json:286]: Without this, the filtering doesn't work

**Comment by sumeetdas-dlpx** [comment]
Closing this PR as we have created a new PR (#24) that consolidates all Azure SQL pipeline code.

**Pull request closed without merging**